lib/cmd/common/admin-devices.js

```diff
@@ -13,108 +12,97 @@ devices.usage = "\nfhc devices list"
 
 var fhreq = require("../../utils/request");
 var common = require("../../common");
+var _ = require('underscore');
+
+var actions = {
+  list: list,
+  read: read,
+  update: update,
+  enable: enabledDevice,
+  purgedata: purgedataDevice,
+  listusers: listusers,
+  listapps: listapps
+};
 
 // Main devices entry point
-function devices (argv, cb) {
+function devices(argv, cb) {
+  /*eslint no-else-return:0 */
   var args = argv._;
-  if (args.length === 0){
+  if (args.length === 0) {
     return list(cb);
   }
 
-  var action = args[0];
-  if (action === 'list') {
-    return list(cb);
-  }else if (action === 'read'){
-    if (args.length !== 2) return cb("Invalid arguments for 'read':" + devices.usage);
-    return read(args[1], cb);
-  }else if (action === 'update'){
-    if (args.length !== 3) return cb("Invalid arguments for 'update':" + devices.usage);
-    return update(args[1], args[2], cb);
-  }else if (action === 'enable'){
-    if (args.length !== 3) return cb("Invalid arguments for 'enable':" + devices.usage);
-    return enabledDevice(args[1], args[2], cb);
-  }else if (action === 'purgedata'){
-    if (args.length !== 3) return cb("Invalid arguments for 'purgedata':" + devices.usage);
-    return purgedataDevice(args[1], args[2], cb);
-  }else if (action === 'listusers'){
-    if (args.length !== 2) return cb("Invalid arguments for 'listusers':" + devices.usage);
-    return listusers(args[1], cb);
-  }else if (action === 'listapps'){
-    if (args.length !== 2) return cb("Invalid arguments for 'listapps':" + devices.usage);
-    return listapps(args[1], cb);
-  }else if (args.length === 1){
-    var deviceId = args[0];
-    return read(deviceId, cb);
-  }else{
+  var action = args.shift();
+  var func = actions[action];
+  if (!func) {
+    // command not found, try `read({deviceId})`
+    if (args.length === 1) {
+      var deviceId = args[0];
+      return read(deviceId, cb);
+    }
     return cb(devices.usage);
   }
+  if (args.length !== func.length + 1) {
+    return cb("Invalid arguments for '" + action + "':" + devices.usage);
+  }
+  return func.command.apply(undefined, args.slice(func.length).concat(cb));
 }
```

Comment:
I'm going to add comments and maybe break down to more statements. `func.length` is actually [Function.length](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Function/length), people just don't remember it exists. Now that I see it is actually `.command` that I forgot to erase and is not defined...

The idea is grabbing the right number of arguments from `args` and keep `cb` as the last one. It's just a tentative design though. I'm going to move it to a separate file and add a specific unit test when I replicate it to the other files that have the same pattern.